File: games/gradius/tools/oracle/input_probe.lua
-- input_probe.lua -- proves the harness can DRIVE the game, not just watch it.
--
-- A recorded playtest is only a permanent test if button presses are injectable
-- and reproducible (docs/knowledge/01, "Corpus"). This runs the same ROM twice
-- in one process is not possible, so it takes a mode from the environment:
--
--   PROBE_PRESS = "none" | "start"
--   PROBE_FRAMES = frame at which the framebuffer is captured
--
-- With "start" it holds START from frame 200 to 209 and releases it. If input
-- injection works, the two modes produce different pictures at the same frame.
-- If it is deterministic, each mode reproduces its own picture exactly.
--
-- NOTE ON TIMING: buttons are applied in the inputPolled event, i.e. at the
-- moment the game actually reads the controller. That is the NES analogue of
-- the Game Boy "input has a lead" rule; do not assume the lead is zero, measure
-- it once for real against a scenario before building the corpus on it.

local function say(s) print("PROBE " .. s) end

local FRAMES = tonumber(os.getenv("PROBE_FRAMES") or "") or 400
local MODE   = os.getenv("PROBE_PRESS") or "none"
local OUT    = os.getenv("PROBE_OUT") or "."

local PRESS_FROM, PRESS_TO = 200, 209

local frame = 0
local polls_seen, polls_forced = 0, 0

emu.addEventCallback(function()
   polls_seen = polls_seen + 1
   if MODE == "start" and frame >= PRESS_FROM and frame <= PRESS_TO then
      emu.setInput({ start = true }, 0)
      polls_forced = polls_forced + 1
   end
end, emu.eventType.inputPolled)

emu.addEventCallback(function()
   frame = frame + 1
   if frame ~= FRAMES then return end

   local fb = emu.getScreenBuffer()
   local h = 2166136261
   local nonblack = 0
   for i = 1, #fb do
      local px = fb[i]
      for _, c in ipairs({ (px >> 16) & 0xFF, (px >> 8) & 0xFF, px & 0xFF }) do
         h = (h ~ c); h = (h * 16777619) & 0xFFFFFFFF
      end
      if (px & 0xFFFFFF) ~= 0 then nonblack = nonblack + 1 end
   end

   local f = assert(io.open(OUT .. "/input_" .. MODE .. ".png", "wb"))
   f:write(emu.takeScreenshot())
   f:close()

   local st = emu.getState()
   say("mode = " .. MODE)
   say("frames.executed = " .. frame)
   say("input.polledEvents = " .. polls_seen)
   say("input.forcedPolls = " .. polls_forced)
   say(("framebuffer.fnv1a = 0x%08X"):format(h))
   say("framebuffer.nonBlackPixels = " .. nonblack)
   say("cpu.cycleCount = " .. tostring(st["cpu.cycleCount"]))
   say("mem.$0000_0003 = " .. string.format("%02X %02X %02X %02X",
       emu.read(0, emu.memType.nesDebug, false), emu.read(1, emu.memType.nesDebug, false),
       emu.read(2, emu.memType.nesDebug, false), emu.read(3, emu.memType.nesDebug, false)))
   say("END")
   emu.stop(0)
end, emu.eventType.endFrame)

File: games/gradius/tools/oracle/input_probe.py
#!/usr/bin/env python3
"""Prove scripted input is injectable AND reproducible.

Runs the same 400 frames four times: {no input, hold START} x {run 1, run 2}.

  - the two modes must produce DIFFERENT pictures  (input actually reaches the game)
  - each mode must reproduce ITSELF byte for byte  (the corpus can be trusted)

A green result that only checked "the run finished" would be worthless here;
that is trap #4.3 - a check with no transition in it cannot fail.

    python games/gradius/tools/oracle/input_probe.py

Outputs under out/input/ are ROM-derived and gitignored.
"""

from __future__ import annotations

import argparse
import hashlib
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import mesen  # noqa: E402

HERE = Path(__file__).resolve().parent
LUA = HERE / "input_probe.lua"


def once(rom: Path, mode: str, frames: int, out: Path) -> tuple[dict, str, int]:
    out.mkdir(parents=True, exist_ok=True)
    run = mesen.run_script(LUA, rom, timeout_s=120, env_extra={
        "PROBE_PRESS": mode, "PROBE_FRAMES": str(frames), "PROBE_OUT": out.as_posix()})
    png = out / f"input_{mode}.png"
    digest = hashlib.sha256(png.read_bytes()).hexdigest() if png.exists() else "MISSING"
    return run.fields(), digest, run.returncode


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--rom", type=Path, default=mesen.DEFAULT_ROM)
    ap.add_argument("--frames", type=int, default=400)
    ap.add_argument("--out", type=Path, default=HERE / "out" / "input")
    args = ap.parse_args()

    results = {}
    for mode in ("none", "start"):
        for rep in (1, 2):
            fields, digest, rc = once(args.rom, mode, args.frames,
                                      args.out / f"{mode}{rep}")
            results[(mode, rep)] = (fields, digest, rc)
            print(f"{mode} run{rep}: exit={rc} "
                  f"fnv1a={fields.get('framebuffer.fnv1a')} "
                  f"forcedPolls={fields.get('input.forcedPolls')} "
                  f"nonBlack={fields.get('framebuffer.nonBlackPixels')} "
                  f"png.sha256={digest[:16]}...")
    print()

    ok = True
    for mode in ("none", "start"):
        a, b = results[(mode, 1)], results[(mode, 2)]
        same = a[0] == b[0] and a[1] == b[1]
        print(f"  [{'PASS' if same else 'FAIL'}] '{mode}' reproduces itself exactly")
        ok = ok and same
    n, s = results[("none", 1)], results[("start", 1)]
    differ = n[1] != s[1]
    print(f"  [{'PASS' if differ else 'FAIL'}] holding START changes the picture "
          f"(input really reaches the game)")
    forced = int(s[0].get("input.forcedPolls", "0"))
    print(f"  [{'PASS' if forced > 0 else 'FAIL'}] setInput ran on {forced} polled frames")
    ok = ok and differ and forced > 0
    print("\nRESULT:", "INPUT INJECTION PROVEN" if ok else "SOMETHING FAILED")
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())
